File: java/core/src/main/java/dev/deja/core/trajectory/Compare.java
package dev.deja.core.trajectory;

import dev.deja.core.cassette.CassetteFrame;
import dev.deja.core.trajectory.Alignment.AlignedPair;
import dev.deja.core.trajectory.Frontier.TrajectorySession;
import lombok.experimental.UtilityClass;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Set;

/** Mirrors the TypeScript implementation's {@code compare.ts}. */
@UtilityClass
public class Compare {

    /** A step is "post-divergence" only once we know an earlier actual request went unserved --
     *  derived from {@link Frontier#detectReplaySession}, never asserted for golden-only
     *  entries. */
    private Phase phaseFor(TrajectoryStep step, TrajectorySession session) {
        if (step == null || session == null || session.firstMissFrameIndex() == null) {
            return null;
        }
        return step.frameIndex() >= session.firstMissFrameIndex() ? Phase.POST_DIVERGENCE : null;
    }

    private final class Counts {
        boolean passed = true;
        int exact;
        int tolerated;
        int drifted;
        int reordered;
        int added;
        int missing;
        int requiredMissing;
        int prohibitedPresent;

        TrajectorySummary toSummary() {
            return new TrajectorySummary(passed, exact, tolerated, drifted, reordered, added, missing, requiredMissing, prohibitedPresent);
        }
    }

    private TrajectoryStepReport classifySequencePair(
            AlignedPair pair, List<TrajectoryStep> golden, List<TrajectoryStep> actual, double threshold, TrajectorySession session) {
        TrajectoryStep goldenStep = pair.goldenIndex() != null ? golden.get(pair.goldenIndex()) : null;
        TrajectoryStep actualStep = pair.actualIndex() != null ? actual.get(pair.actualIndex()) : null;
        TrajectoryStep step = goldenStep != null ? goldenStep : actualStep;

        StepOutcome outcome;
        if (goldenStep != null && actualStep != null) {
            outcome = pair.reordered()
                    ? StepOutcome.REORDERED
                    : pair.score() != null && pair.score() == 1
                            ? StepOutcome.EXACT
                            : (pair.score() != null ? pair.score() : 0) >= threshold ? StepOutcome.TOLERATED : StepOutcome.DRIFTED;
        } else {
            outcome = goldenStep != null ? StepOutcome.MISSING : StepOutcome.ADDED;
        }

        return new TrajectoryStepReport(
                goldenStep != null ? goldenStep.index() : null,
                actualStep != null ? actualStep.index() : null,
                outcome,
                step.method(),
                step.toolName(),
                pair.score(),
                pair.tier(),
                phaseFor(actualStep, session));
    }

    /** {@code drifted} always fails; whether an unmatched side fails depends on mode (subset
     *  tolerates unmatched golden, superset tolerates unmatched actual, strict/unordered
     *  tolerate neither). */
    private boolean isSequenceFailure(StepOutcome outcome, TrajectoryMode mode) {
        // Still a failure: the label improves diagnosis only, it never changes gate semantics.
        // annotateReorders only ever fires in strict mode, where this was already going to
        // fail as disconnected missing/added entries.
        if (outcome == StepOutcome.DRIFTED || outcome == StepOutcome.REORDERED) {
            return true;
        }
        if (outcome == StepOutcome.MISSING) {
            return mode != TrajectoryMode.SUBSET;
        }
        if (outcome == StepOutcome.ADDED) {
            return mode != TrajectoryMode.SUPERSET;
        }
        return false;
    }

    private TrajectoryReport buildSequenceReport(
            TrajectoryMode mode, double threshold, List<TrajectoryStep> golden, List<TrajectoryStep> actual,
            List<AlignedPair> pairs, TrajectorySession session) {
        Counts counts = new Counts();
        List<TrajectoryStepReport> steps = new ArrayList<>();

        for (AlignedPair pair : pairs) {
            TrajectoryStepReport report = classifySequencePair(pair, golden, actual, threshold, session);
            switch (report.outcome()) {
                case EXACT -> counts.exact++;
                case TOLERATED -> counts.tolerated++;
                case DRIFTED -> counts.drifted++;
                case REORDERED -> counts.reordered++;
                case MISSING -> counts.missing++;
                case ADDED -> counts.added++;
                default -> { }
            }
            if (isSequenceFailure(report.outcome(), mode)) {
                counts.passed = false;
            }
            steps.add(report);
        }

        return new TrajectoryReport(TrajectoryReport.CURRENT_VERSION, mode, threshold, steps, session, counts.toSummary());
    }

    private TrajectoryReport evaluatePolicy(List<TrajectoryStep> actual, TrajectoryPolicy policy, TrajectorySession session) {
        Counts counts = new Counts();
        List<TrajectoryStepReport> steps = new ArrayList<>();

        for (PolicyStepMatcher matcher : policy.required()) {
            Optional<TrajectoryStep> found = Policy.findPolicyMatch(matcher, actual);
            if (found.isPresent()) {
                steps.add(new TrajectoryStepReport(
                        null, found.get().index(), StepOutcome.REQUIRED_SATISFIED,
                        matcher.method(), matcher.toolName(), null, null, phaseFor(found.get(), session)));
            } else {
                counts.requiredMissing++;
                counts.passed = false;
                steps.add(new TrajectoryStepReport(null, null, StepOutcome.REQUIRED_MISSING, matcher.method(), matcher.toolName(), null, null, null));
            }
        }

        for (PolicyStepMatcher matcher : policy.prohibited()) {
            for (TrajectoryStep found : Policy.findAllPolicyMatches(matcher, actual)) {
                counts.prohibitedPresent++;
                counts.passed = false;
                steps.add(new TrajectoryStepReport(
                        null, found.index(), StepOutcome.PROHIBITED_PRESENT,
                        matcher.method(), matcher.toolName(), null, null, phaseFor(found, session)));
            }
        }

        for (PolicyStepMatcher matcher : policy.optional()) {
            Optional<TrajectoryStep> found = Policy.findPolicyMatch(matcher, actual);
            if (found.isPresent()) {
                steps.add(new TrajectoryStepReport(
                        null, found.get().index(), StepOutcome.OPTIONAL_OBSERVED,
                        matcher.method(), matcher.toolName(), null, null, phaseFor(found.get(), session)));
            } else {
                steps.add(new TrajectoryStepReport(null, null, StepOutcome.OPTIONAL_ABSENT, matcher.method(), matcher.toolName(), null, null, null));
            }
        }

        if (policy.closedWorld()) {
            List<PolicyStepMatcher> allMatchers = new ArrayList<>();
            allMatchers.addAll(policy.required());
            allMatchers.addAll(policy.prohibited());
            allMatchers.addAll(policy.optional());

            for (TrajectoryStep step : actual) {
                if (Policy.isCoveredByAnyMatcher(step, allMatchers)) {
                    continue;
                }
                counts.prohibitedPresent++;
                counts.passed = false;
                steps.add(new TrajectoryStepReport(
                        null, step.index(), StepOutcome.PROHIBITED_PRESENT,
                        step.method(), step.toolName(), null, null, phaseFor(step, session)));
            }
        }

        return new TrajectoryReport(TrajectoryReport.CURRENT_VERSION, TrajectoryMode.POLICY, CompareOptions.DEFAULT_THRESHOLD, steps, session, counts.toSummary());
    }

    /** Above this, {@code unordered}/{@code subset}/{@code superset} mode's per-group Hungarian
     *  assignment (O(n<sup>3</sup>) in the size of the *largest same-(method,toolName) group*,
     *  not the trajectory as a whole) risks hanging on a pathological input rather than failing
     *  fast. Real trajectories -- an agent's actual tool calls in one run -- are tens to low
     *  hundreds of steps; this is headroom, not a realistic ceiling anyone should ever hit. */
    private static final int MAX_TRAJECTORY_STEPS = 5000;

    private static void checkTrajectorySize(List<TrajectoryStep> golden, List<TrajectoryStep> actual) {
        if (golden.size() > MAX_TRAJECTORY_STEPS || actual.size() > MAX_TRAJECTORY_STEPS) {
            throw new IllegalArgumentException("Deja: trajectory too large to compare (golden " + golden.size()
                    + " steps, actual " + actual.size() + " steps, max " + MAX_TRAJECTORY_STEPS + " per side)");
        }
    }

    /** Compares two already-extracted trajectories. {@code session} -- if the caller has
     *  divergence-frontier evidence (see {@link Frontier#detectReplaySession}) -- tags
     *  post-divergence steps. */
    public TrajectoryReport compareTrajectories(
            List<TrajectoryStep> golden, List<TrajectoryStep> actual, CompareOptions options, TrajectorySession session) {
        checkTrajectorySize(golden, actual);
        TrajectoryMode mode = options.modeOrDefault();
        double threshold = options.thresholdOrDefault();

        if (mode == TrajectoryMode.POLICY) {
            if (options.policy() == null) {
                throw new IllegalArgumentException("policy mode requires a policy document (see --policy / --derive-policy)");
            }
            return evaluatePolicy(actual, options.policy(), session);
        }

        List<AlignedPair> pairs = mode == TrajectoryMode.STRICT
                ? Alignment.annotateReorders(Alignment.alignStrict(golden, actual, threshold), golden, actual, threshold)
                : Alignment.alignUnordered(golden, actual, threshold);
        return buildSequenceReport(mode, threshold, golden, actual, pairs, session);
    }

    public TrajectoryReport compareTrajectories(List<TrajectoryStep> golden, List<TrajectoryStep> actual, CompareOptions options) {
        return compareTrajectories(golden, actual, options, null);
    }

    /** Convenience entry point mirroring the CLI: extracts both trajectories from raw wire
     *  frames and recovers divergence-frontier evidence from the actual cassette itself. */
    public TrajectoryReport compareCassetteFrames(
            List<CassetteFrame> goldenFrames, List<CassetteFrame> actualFrames, CompareOptions options, Set<String> include) {
        List<TrajectoryStep> golden = Extract.extractTrajectory(goldenFrames, include);
        List<TrajectoryStep> actual = Extract.extractTrajectory(actualFrames, include);
        TrajectorySession session = Frontier.detectReplaySession(actualFrames).orElse(null);
        return compareTrajectories(golden, actual, options, session);
    }

    public TrajectoryReport compareCassetteFrames(List<CassetteFrame> goldenFrames, List<CassetteFrame> actualFrames, CompareOptions options) {
        return compareCassetteFrames(goldenFrames, actualFrames, options, Set.of());
    }
}

File: ts/src/trajectory/compare.ts
import { CassetteFrame } from "../core/types.js";
import { AlignedPair, alignStrict, alignUnordered, annotateReorders } from "./alignment.js";
import { extractTrajectory } from "./extract.js";
import { detectReplaySession } from "./frontier.js";
import { derivePolicy, findAllPolicyMatches, findPolicyMatch, isCoveredByAnyMatcher } from "./policy.js";
import {
    StepOutcome,
    TrajectoryCompareOptions,
    TrajectoryMode,
    TrajectoryPolicy,
    TrajectoryReport,
    TrajectorySession,
    TrajectoryStep,
    TrajectoryStepReport,
    TrajectorySummary,
} from "./types.js";

const DEFAULT_THRESHOLD = 0.75;

function emptySummary(): TrajectorySummary {
    return {
        passed: true,
        exact: 0,
        tolerated: 0,
        drifted: 0,
        reordered: 0,
        added: 0,
        missing: 0,
        requiredMissing: 0,
        prohibitedPresent: 0,
    };
}

/** A step is "post-divergence" only once we know an earlier actual request went unserved --
 *  derived from `detectReplaySession`, never asserted for golden-only entries. */
function phaseFor(step: TrajectoryStep | undefined, session: TrajectorySession | undefined): "post-divergence" | undefined {
    if (!step || session?.firstMissFrameIndex === undefined) return undefined;
    return step.frameIndex >= session.firstMissFrameIndex ? "post-divergence" : undefined;
}

function classifySequencePair(
    pair: AlignedPair,
    golden: TrajectoryStep[],
    actual: TrajectoryStep[],
    threshold: number,
    session: TrajectorySession | undefined
): TrajectoryStepReport {
    const goldenStep = pair.goldenIndex !== undefined ? golden[pair.goldenIndex] : undefined;
    const actualStep = pair.actualIndex !== undefined ? actual[pair.actualIndex] : undefined;
    const step = (goldenStep ?? actualStep)!;

    let outcome: StepOutcome;
    if (goldenStep && actualStep) {
        outcome = pair.reordered
            ? "reordered"
            : pair.score === 1
              ? "exact"
              : (pair.score ?? 0) >= threshold
                ? "tolerated"
                : "drifted";
    } else {
        outcome = goldenStep ? "missing" : "added";
    }

    return {
        goldenIndex: goldenStep?.index,
        actualIndex: actualStep?.index,
        outcome,
        method: step.method,
        toolName: step.toolName,
        score: pair.score,
        tier: pair.tier,
        phase: phaseFor(actualStep, session),
    };
}

/** `drifted` always fails; whether an unmatched side fails depends on mode (subset tolerates
 *  unmatched golden, superset tolerates unmatched actual, strict/unordered tolerate neither). */
function isSequenceFailure(outcome: StepOutcome, mode: TrajectoryMode): boolean {
    // Still a failure: the label improves diagnosis only, it never changes gate semantics.
    // annotateReorders only ever fires in strict mode, where this was already going to fail
    // as disconnected missing/added entries.
    if (outcome === "drifted" || outcome === "reordered") return true;
    if (outcome === "missing") return mode !== "subset";
    if (outcome === "added") return mode !== "superset";
    return false;
}

function buildSequenceReport(
    mode: TrajectoryMode,
    threshold: number,
    golden: TrajectoryStep[],
    actual: TrajectoryStep[],
    pairs: AlignedPair[],
    session: TrajectorySession | undefined
): TrajectoryReport {
    const summary = emptySummary();

    const steps = pairs.map((pair) => {
        const report = classifySequencePair(pair, golden, actual, threshold, session);
        switch (report.outcome) {
            case "exact":
                summary.exact++;
                break;
            case "tolerated":
                summary.tolerated++;
                break;
            case "drifted":
                summary.drifted++;
                break;
            case "reordered":
                summary.reordered++;
                break;
            case "missing":
                summary.missing++;
                break;
            case "added":
                summary.added++;
                break;
        }
        if (isSequenceFailure(report.outcome, mode)) summary.passed = false;
        return report;
    });

    return { reportVersion: 1, mode, threshold, steps, session, summary };
}

function evaluatePolicy(
    actual: TrajectoryStep[],
    policy: TrajectoryPolicy,
    session: TrajectorySession | undefined
): TrajectoryReport {
    const summary = emptySummary();
    const steps: TrajectoryStepReport[] = [];

    for (const matcher of policy.required ?? []) {
        const found = findPolicyMatch(matcher, actual);
        if (found) {
            steps.push({
                actualIndex: found.index,
                outcome: "required-satisfied",
                method: matcher.method,
                toolName: matcher.toolName,
                phase: phaseFor(found, session),
            });
        } else {
            summary.requiredMissing++;
            summary.passed = false;
            steps.push({ outcome: "required-missing", method: matcher.method, toolName: matcher.toolName });
        }
    }

    for (const matcher of policy.prohibited ?? []) {
        for (const found of findAllPolicyMatches(matcher, actual)) {
            summary.prohibitedPresent++;
            summary.passed = false;
            steps.push({
                actualIndex: found.index,
                outcome: "prohibited-present",
                method: matcher.method,
                toolName: matcher.toolName,
                phase: phaseFor(found, session),
            });
        }
    }

    for (const matcher of policy.optional ?? []) {
        const found = findPolicyMatch(matcher, actual);
        steps.push(
            found
                ? {
                      actualIndex: found.index,
                      outcome: "optional-observed",
                      method: matcher.method,
                      toolName: matcher.toolName,
                      phase: phaseFor(found, session),
                  }
                : { outcome: "optional-absent", method: matcher.method, toolName: matcher.toolName }
        );
    }

    if (policy.closedWorld) {
        const allMatchers = [...(policy.required ?? []), ...(policy.prohibited ?? []), ...(policy.optional ?? [])];
        for (const step of actual) {
            if (isCoveredByAnyMatcher(step, allMatchers)) continue;
            summary.prohibitedPresent++;
            summary.passed = false;
            steps.push({
                actualIndex: step.index,
                outcome: "prohibited-present",
                method: step.method,
                toolName: step.toolName,
                phase: phaseFor(step, session),
            });
        }
    }

    return { reportVersion: 1, mode: "policy", threshold: DEFAULT_THRESHOLD, steps, session, summary };
}

/** Above this, `unordered`/`subset`/`superset` mode's per-group Hungarian assignment (O(n³) in
 *  the size of the *largest same-(method,toolName) group*, not the trajectory as a whole) risks
 *  hanging on a pathological input rather than failing fast. Real trajectories -- an agent's
 *  actual tool calls in one run -- are tens to low hundreds of steps; this is headroom, not a
 *  realistic ceiling anyone should ever hit. */
const MAX_TRAJECTORY_STEPS = 5000;

function checkTrajectorySize(golden: TrajectoryStep[], actual: TrajectoryStep[]): void {
    if (golden.length > MAX_TRAJECTORY_STEPS || actual.length > MAX_TRAJECTORY_STEPS) {
        throw new Error(
            `Deja: trajectory too large to compare (golden ${golden.length} steps, actual ${actual.length} steps, max ${MAX_TRAJECTORY_STEPS} per side)`
        );
    }
}

/** Compares two already-extracted trajectories. `session` -- if the caller has
 *  divergence-frontier evidence (see `detectReplaySession`) -- tags post-divergence steps. */
export function compareTrajectories(
    golden: TrajectoryStep[],
    actual: TrajectoryStep[],
    options: TrajectoryCompareOptions = {},
    session?: TrajectorySession
): TrajectoryReport {
    checkTrajectorySize(golden, actual);
    const mode = options.mode ?? "strict";
    const threshold = options.threshold ?? DEFAULT_THRESHOLD;

    if (mode === "policy") {
        if (!options.policy) throw new Error("policy mode requires a policy document (see --policy / --derive-policy)");
        return evaluatePolicy(actual, options.policy, session);
    }

    const pairs =
        mode === "strict"
            ? annotateReorders(alignStrict(golden, actual, threshold), golden, actual, threshold)
            : alignUnordered(golden, actual, threshold);
    return buildSequenceReport(mode, threshold, golden, actual, pairs, session);
}

export interface CompareCassetteFramesOptions extends TrajectoryCompareOptions {
    /** Extraction inclusion overrides (`--include`) -- distinct from the `policy`
     *  comparison mode's `TrajectoryPolicy` despite the naming collision. */
    include?: string[];
}

/** Convenience entry point for the `deja trajectory` CLI: extracts both trajectories from raw
 *  wire frames and recovers divergence-frontier evidence from the actual cassette itself. */
export function compareCassetteFrames(
    goldenFrames: CassetteFrame[],
    actualFrames: CassetteFrame[],
    options: CompareCassetteFramesOptions = {}
): TrajectoryReport {
    const golden = extractTrajectory(goldenFrames, { include: options.include });
    const actual = extractTrajectory(actualFrames, { include: options.include });
    const session = detectReplaySession(actualFrames);
    return compareTrajectories(golden, actual, options, session);
}

export { derivePolicy };
